### PR TITLE
[WIP] Integrating libp2p into Lodestar

### PR DESCRIPTION
**Pull request by Mikerah**
This is currently a work in progress and breaks tests.

I have started integrating libp2p into Lodestar. A simple libp2p bundle has been created and is subject to change as more components get built out. 

Here's a current TODO list:

- [ ] Have a working basic server for managing peers

- [ ] (Optional) Change part of the service API

- [ ] Add RPC over libp2p to start integrating the ETH2.0 Wire Protocol

- [ ] Add BDD tests to make sure networking is fine.

### CODE REVIEW

**Comment by GregTheGreek** [comment]
Just quickly glanced, we should be probably importing a libp2p library rather than a bunch of individual ones?

**Comment by Mikerah** [comment]
@GregTheGreek The way it works is that if you don't need any specialized transports, multiplexers, etc and you can use the ones provided by libp2p, you need to add them to the project individually. The one that might be optional is FloodSub but I still imported so that when GossipSub is done, I can just replace the import.

File: src/p2p/index.ts
Comment (wemeetagain) [inline src/p2p/index.ts]: #156 adds a logger, we'll probably want to use that here eventually.

File: src/p2p/index.ts
Comment (wemeetagain) [inline src/p2p/index.ts]: we've got a the `promisify-es6` package that can help smooth this pattern over and convert callback-based code into promise-based code.
eg: `await promisify(this.node.start.bind(this.node))();`

File: src/p2p/node.ts
Comment (wemeetagain) [inline src/p2p/node.ts]: You might think about renaming this to highlight that this is a more Nodejs-specific config vs a browser-specific config.

File: src/p2p/node.ts
Comment (wemeetagain) [inline src/p2p/node.ts]: consider promisifying here vs using waterfall
eg: `const peerInfo = await promisify(PeerInfo.create)(peerId);`

File: src/p2p/index.ts
Comment (wemeetagain) [inline src/p2p/index.ts]: this can probably go in the constructor

**Review by wemeetagain — CHANGES_REQUESTED** [review]
very exciting, starting to take shape :)

**Comment by wemeetagain** [comment]
@GregTheGreek I think this is as it should be re: importing libraries for now. libp2p takes a veeery modular approach, and we need all of these libraries to weave everything together properly. In fact, we probably will be adding even more libp2p libraries in the future!

Its probably easier to start here, monorepo style, and once we've got things more solid, pull out into a separate repo/ repos, if it makes sense.

File: src/p2p/node.ts
Comment (Mikerah) [reply src/p2p/node.ts]: Indeed. I have just been working on a nodejs specific libp2p bundle. The browser specific bundle should mostly be the same except the transport, etc is different

**Comment by GregTheGreek** [comment]
@wemeetagain Fair enough, I iguess we have a lot to trimfrom the package.json anyways.

**Comment by wemeetagain** [comment]
@Mikerah would you be willing to do minor cleanup and plan on merging as is?
It would probably be better to just have the code in master and iterate from there.
I think your TODOs could be reintroduced as issues for further PRs.

**Comment by Mikerah** [comment]
@wemeetagain I can do minor cleanups and separate the TODO as separate PRs. My main bottleneck right now is that all the tests break due to odd compilation issues. Before merging, I would like to confirm that this is just an issue on my end and not everybody else's.

**Comment by wemeetagain** [comment]
I think what you're seeing is a result of our `check-types` script.
Before running tests, we run the typescript compiler (without output) just for it to do some checks and catch any type errors.
It's very picky and will catch many small issues.
You'd probably want to just go through them, one by one.

If you're using an IDE (or vim w/ plugins like [typescript-vim](https://github.com/leafgarland/typescript-vim) and [YouCompleteMe](https://github.com/Valloric/YouCompleteMe)), it may be much easier, because the errors will be called out in the editor and potentially can be auto-fixed.

**Comment by Mikerah** [comment]
The basics are in here and I will be opening subsequent PRs. @mpetrunic got the types working.

File: package.json
Comment (mpetrunic) [inline package.json]: I think this should be dev dependency

Comment (mpetrunic) [reply package.json]: also whats up with two ssz libs

Comment (Mikerah) [reply package.json]: I think we can safely remove the second ssz dependency. There was an org change in the name and that's why that is there.

File: package.json
Comment (mpetrunic) [inline package.json:88]: oh there was misunderstanding, I meant to move `@polakdot/ts` only because those are typings that aren't used once the library is built.

File: src/p2p/index.ts
Comment (GregTheGreek) [reply src/p2p/index.ts]: can we update this?

**Review by GregTheGreek — DISMISSED** [review]
LGTM -- Can we fix the logger?

File: package.json
Comment (wemeetagain) [inline package.json]: old dependency

File: src/p2p/index.ts
Comment (wemeetagain) [inline src/p2p/index.ts]: can you promisify like in `start`?

File: src/p2p/index.ts
Comment (wemeetagain) [inline src/p2p/index.ts]: We should just `await` this, instead of storing in a variable.

File: src/p2p/node.ts
Comment (wemeetagain) [inline src/p2p/node.ts]: We aren't using this, do we need it?

File: src/p2p/index.ts
Comment (wemeetagain) [inline src/p2p/index.ts:116]: We should remove all listeners from this.node here, eg: `this.node.removeAllListeners()`

**Review by wemeetagain — CHANGES_REQUESTED** [review]

File: src/p2p/node.ts
Comment (Mikerah) [reply src/p2p/node.ts]: The idea here was to have a singleton node. I still haven't pushed these changes remotely.

**Comment by Mikerah** [comment]
I have pushed some new changes. I turned the libp2p bundle into a singleton and some other stuff.

**Comment by wemeetagain** [comment]
LGTM :+1: 
github says there are some file conflicts, will add approval after that

**Review by wemeetagain — DISMISSED** [review]

**Comment by GregTheGreek** [comment]
The CI is failing on the linter, can you try running `yarn run lint-fix`

**Comment by wemeetagain** [comment]
you'll need to fix the linter and typescript errors before we can merge, but they're all really minor things.

linter errors
```
/home/travis/build/ChainSafe/lodestar/src/p2p/index.ts
  146:1   error    Expected indentation of 8 spaces but found 1 tab  @typescript-eslint/indent
  148:1   error    Expected indentation of 8 spaces but found 1 tab  @typescript-eslint/indent

/home/travis/build/ChainSafe/lodestar/src/p2p/node.ts
  41:1  error  Expected indentation of 2 spaces but found 4    @typescript-eslint/indent
  47:1  error  Expected indentation of 8 spaces but found 9    @typescript-eslint/indent
  48:1  error  Expected indentation of 8 spaces but found 9    @typescript-eslint/indent
  49:1  error  Expected indentation of 10 spaces but found 11  @typescript-eslint/indent
  50:1  error  Expected indentation of 8 spaces but found 9    @typescript-eslint/indent
  51:1  error  Expected indentation of 8 spaces but found 9    @typescript-eslint/indent
  52:1  error  Expected indentation of 6 spaces but found 8    @typescript-eslint/indent
```

typescript errors
```
src/cli/commands/beacon.ts(30,33): error TS2345: Argument of type '{ db: { name: any; }; eth1: { depositContract: { address: any; }; provider: BaseProvider; }; }' is not assignable to parameter of type 'BeaconNodeCtx'.
  Type '{ db: { name: any; }; eth1: { depositContract: { address: any; }; provider: BaseProvider; }; }' is missing the following properties from type 'BeaconNodeCtx': chain, network, rpc, sync, opPool
src/node/index.ts(54,40): error TS2345: Argument of type 'object' is not assignable to parameter of type 'EthersEth1Options'.
  Type '{}' is missing the following properties from type 'EthersEth1Options': provider, depositContract
src/p2p/index.ts(4,9): error TS2305: Module '"../logger"' has no exported member 'logger'.
src/p2p/index.ts(45,11): error TS2554: Expected 0 arguments, but got 1.
src/p2p/node.ts(41,47): error TS1064: The return type of an async function or method must be the global Promise<T> type.
```
Re BeaconNodeCtx, you will probably need to let the interface properties be optional for now and also change the `eth1` property back to an `any`. We can address the typing in a later PR, but that should fix that problem for now.
Re logger, you'll need to import the default object, eg: `import logger from "../logger";`
Re expected 0 arguments, you should remove the `super` since there's not a superclass we're interested in.
Re return type of async, `createNode` shouldn't be declared as an `async` function since its using callback style

**Comment by Mikerah** [comment]
The build is now passing. Awaiting approval...

**Review by wemeetagain — DISMISSED** [review]

File: src/p2p/index.ts
Comment (GregTheGreek) [inline src/p2p/index.ts]: I'm not sure this needs to be async

File: src/p2p/index.ts
Comment (GregTheGreek) [inline src/p2p/index.ts]: as well as this

File: src/p2p/index.ts
Comment (GregTheGreek) [inline src/p2p/index.ts]: this as well

File: src/p2p/node.ts
Comment (GregTheGreek) [reply src/p2p/node.ts]: I think using promisify may help remove some of the callback hell

**Review by GregTheGreek — CHANGES_REQUESTED** [review]
Few small async based comments, but also theres a lot of callback hell going on, using `promisify` would really help

**Comment by Mikerah** [comment]
@GregTheGreek I think leaving the `waterfall` in `createNode` is fine. This function won't be changing much in the future. I also think that `waterfall` is more readable than `promisify` in the way it's currently coded.  However, I found [this issue](https://github.com/caolan/async/issues/956) that seems like it's a compromise between both ways.

**Comment by mpetrunic** [comment]
> @GregTheGreek I think leaving the `waterfall` in `createNode` is fine. This function won't be changing much in the future. I also think that `waterfall` is more readable than `promisify` in the way it's currently coded. However, I found [this issue](https://github.com/caolan/async/issues/956) that seems like it's a compromise between both ways.

I think the point is we don't wan't both async an promisify librarys since they both do same job

**Comment by wemeetagain** [comment]
It looks like `createNode` was pulled from floodsub.
In gossipsub, I recently promisifyed it, you can see it here: https://github.com/ChainSafe/gossipsub-js/blob/gossip_implementation/test/utils.js#L18

**Comment by Mikerah** [comment]
@wemeetagain `createNode` was pulled from one of the `libp2p` examples. They use that pretty much everywhere. I don't mind promisifying.

**Comment by wemeetagain** [comment]
You can remove the `async` dependency

Edit: you should take a look at the typescript checking too (`yarn check-types`), causing travis failure.

**Review by wemeetagain — APPROVED** [review]

File: src/p2p/node.ts
Comment (GregTheGreek) [inline src/p2p/node.ts]: You can remove the `let` on line 43 and replace line 48 with the assignment
`const node: LodestarNode = new LodestarNode({ ... })`

**Review by GregTheGreek — CHANGES_REQUESTED** [review]

File: src/p2p/node.ts
Comment (wemeetagain) [inline src/p2p/node.ts]: I didn't see these start calls. They shouldn't be in `createNode`, because we want to start nodes separate from creation.

Really we should make our own `start` method thats something like:
```typescript
async start() {
  await promisify(super.start.bind(this))()
  await promisify(this.pubsub.start.bind(this.pubsub))()
}
```

**Review by wemeetagain — CHANGES_REQUESTED** [review]

**Review by wemeetagain — APPROVED** [review]

**Review by GregTheGreek — APPROVED** [review]

**Pull request merged**
